IPP feedback banner business logic

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->
#### Business logic for IPP feedback survey banner

**Note:** This PR is part of the IPP in-app feedback banner project https://github.com/woocommerce/woocommerce-android/issues/8116. It targets the feature branch feature/8116-ipp-feedback-banner.

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR introduces two use case classes containing the business logic of the IPP feedback banner's behavior.

* `ShouldShowFeedbackBanner` contains logic deciding if the banner should be shown.
* `GetIPPFeedbackSurveyUrl` returns the survey URL dedicated to the current user.

In the next PR, I'll add the next use cases responsible for storing banner interactions (actions) in SharedPreferences.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

There is no UI part associated with the use cases yet, so I just recommend to review unit tests: `ShouldShowFeedbackBannerTest` and `GetIPPFeedbackSurveyUrlTest`, and make sure they are complete.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->


- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Extract logic of payments plugin detection

samiuelson committed:
Implement use case for deciding if banner should be displayed

samiuelson committed:
Format line length

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/GetActivePaymentsPlugin.kt:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Use case to get the active payments plugin.
+ */
+class GetActivePaymentsPlugin @Inject constructor(
```
This piece of code is extracted from `PaymentChargeRepository` and reused in `ShouldShowFeedbackBanner` class.

samiuelson committed:
Implement logic returning survey url for user group

samiuelson committed:
Clean up code in `ShouldShowFeedbackBanner`

samiuelson committed:
Clean up code in `GetIPPFeedbackSurveyUrl`

samiuelson committed:
Bump FluxC version

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackSurveyUrl.kt:
```diff
@@ -0,0 +1,73 @@
+    private val getActivePaymentsPlugin: GetActivePaymentsPlugin,
+) {
+    @Suppress("ReturnCount")
+    @Throws(IllegalStateException::class)
```
❓ I think the @throws annotation here isn't very useful since I believe the class which invokes this method will be a Kotlin class and the caller of this method won't be forced to catch this error in the Kotlin file. I feel the caller of this method will not be having knowledge of the exception it throws unless they open this class and reads the code.

AFAIU, this annotation is useful only if we expect this method to be invoked from the Java code. Please correct me if I'm wrong.

AnirudhBhat replied:
Maybe we can model the result such that it returns either the survey URL or the failure details (Why the survey URL couldn't be returned). This way, we let the caller of the method decide how to proceed further with the information. Just a thought.

samiuelson replied:
Good catch! We won't use it in Java code probably.

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackSurveyUrl.kt:
```diff
@@ -0,0 +1,73 @@
+            timeWindowStartDate.formatToYYYYmmDD()
+        )
+
+        if (response.isError || response.result == null) return null
```
❓ I think we might need the failure cause if we decide to log the failure event. In that case, we need to return the failure cause instead of null. WDYT?

samiuelson replied:
Good idea!

AnirudhBhat commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackSurveyUrlTest.kt:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    @Test
+    fun `given no active payments plugin found, should throw exception`() = runBlocking {
```
💡 We could write a similar test to verify that the `numberOfTransactions` must be positive else throw `IllegalStateException`

samiuelson replied:
Nice idea!

AnirudhBhat commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackSurveyUrlTest.kt:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    @Test
+    fun `given banner should not be displayed, should throw exception`() = runBlocking {
```
🔍 Super nitpick: We usually follow the `given, when, then` pattern. We could maybe write the test name as `given banner should not be displayed, then throw exception`? How does this sound? Similarly for other test names.

samiuelson replied:
Thanks for pointing this out!

AnirudhBhat reviewed (commented):
Excellent work on this @samiuelson! I've left a few minor comments to think about. Let me know what you think about those.

AnirudhBhat commented:
@samiuelson Just wanted to mention that all the comments are non-blocking. I just want to listen to your thoughts before we merge.

samiuelson committed:
Simplify `GetActivePaymentsPlugin::invoke` signature

Declare function as operator function.

samiuelson committed:
Improve `GetIPPFeedbackSurveyUrl` use case

1. Use case returns not non-null result, throwing exceptions when API returns error or null data.
2. Added additional unit test to verify behavior in case of negative transactions count.
3. Cleaned up the code around unit tests.

samiuelson committed:
Simplify use case invocations

samiuelson commented:
Thanks, @AnirudhBhat for the in-depth review! I've made the changes.

samiuelson committed:
Revert file added by mistake
